### PR TITLE
RSDK-2537 Instrument gostream with image reading and encoding times

### PR DESCRIPTION
**Pull request by RyanWallon**
This PR instruments functions relating to image reading and encoding to record their execution time.

### CODE REVIEW

**Comment by bazile-clyde** [comment]
@RyanBoring do you have logs in GCP's explorer for what this looks like?

**Comment by RyanWallon** [comment]
[Here's one trace](https://console.cloud.google.com/traces/list?project=av-canary&tid=5794368834d2546ff575ddbd6e4a353f) I got from modifying the media_test.go to start a new exporter.

File: media.go [hunk 1]
```diff
@@ -468,8 +505,9 @@ func (ms *mediaSource[T, U]) Stream(ctx context.Context, errHandlers ...ErrorHan
 		producerCond := sync.NewCond(condMu)
 		consumerCond := sync.NewCond(condMu.RLocker())
 
+		rootCancelCtxWithSpan := trace.NewContext(ms.rootCancelCtx, trace.FromContext(ctx))
```
Comment (bazile-clyde) [inline media.go]: `trace.FromContext` may return `nil` if there's no span. Would that cause a panic in `trace.NewContext`?

File: media.go [hunk 2]
```diff
@@ -190,6 +198,8 @@ func newMediaSource[T, U any](d driver.Driver, r MediaReader[T], p U) MediaSourc
 }
 
 func (pc *producerConsumer[T, U]) start() {
+	startLocalCtx, span := trace.StartSpan(pc.rootCancelCtx, "gostream::producerConsumer::start")
```
Comment (bazile-clyde) [inline media.go]: So here we're creating a span that may not call `span.End()`. This function will only spawn a new goroutine iff `pc.listeners == 1`. You can test this by opening up multiple client streams and streaming at the same time and printing out the value of `pc.listeners`. It should sometimes be `2`. Only one of those will call `defer span.End()`. The other will create the span with this line and return from line 209. You could probably fix this just by calling `span.End()` before the return on line 209. Normally you'd defer but that doesn't look as practical here. The few lines between here and the return will most likely never panic.

File: media.go [hunk 3]
```diff
@@ -239,6 +255,8 @@ func (pc *producerConsumer[T, U]) start() {
 				return
 			}
 
+			var doReadSpan *trace.Span
+			startLocalCtx, doReadSpan = trace.StartSpan(startLocalCtx, "gostream::producerConsumer (anonymous function to read)")
```
Comment (bazile-clyde) [inline media.go]: Why not do this in the anonymous function below and add `doReadSpan.End()` to the defer func in that same function after line 265?

**Comment by bazile-clyde** [comment]
@RyanBoring do you know how well this works with your distributed logging fix or is it too early to tell given that code isn't merged yet?

**Comment by RyanWallon** [comment]
@bazile-clyde I can't really say, I think I'd have to write a client/server that remotely runs some of that code to engage the interceptors. I tried to retrofit the EchoServer in goutils to just run media_test.go from gostream, but since it uses gostream and gostream uses goutils, Go is complaining about an import cycle. I can look into it more tomorrow

File: media.go (hunk 1 above)
Comment (RyanWallon) [reply media.go]: Calling `trace.NewContext` with a `nil` span would append a `nil` span to the context which would later be handled gracefully if it was ever accessed, but it would be unnecessary to leave in there. The new changes don't add the span if it's `nil`.

File: media.go (hunk 2 above)
Comment (RyanWallon) [reply media.go]: That's a good catch!

File: media.go (hunk 3 above)
Comment (RyanWallon) [reply media.go]: Also definitely an oversight on my part!

**Comment by RyanWallon** [comment]
@bazile-clyde I ended up making it work - [here's the trace generated by that test](https://console.cloud.google.com/traces/list?project=av-canary&tid=57c09d1328b9ec7cbf28a2fde55df44b&pageState=(%22traceIntervalPicker%22:(%22groupValue%22:%22PT6H%22,%22customValue%22:null))&start=1689079270980&end=1689100870980). You can see that the client calls the server which runs the media_test.go test with its instrumentation.

File: codec/x264/encoder.go [hunk 4]
```diff
@@ -56,7 +57,10 @@ func (v *encoder) Read() (img image.Image, release func(), err error) {
 }
 
 // Encode asks the codec to process the given image.
-func (v *encoder) Encode(_ context.Context, img image.Image) ([]byte, error) {
+func (v *encoder) Encode(ctx context.Context, img image.Image) ([]byte, error) {
+	_, span := trace.StartSpan(ctx, "x264::Encode")
+	defer span.End()
```
Comment (bazile-clyde) [inline codec/x264/encoder.go]: I'd hold off on adding this until we have some manual tests. That way we won't get a false sense of confidence about what does and does not work. 

Same for line 381

File: media.go [hunk 5]
```diff
@@ -190,25 +198,33 @@ func newMediaSource[T, U any](d driver.Driver, r MediaReader[T], p U) MediaSourc
 }
 
 func (pc *producerConsumer[T, U]) start() {
+	startLocalCtx, span := trace.StartSpan(pc.cancelCtx, "gostream::producerConsumer::start")
+
```
Comment (bazile-clyde) [inline media.go:202]: Nice! Yeah, the results look exactly as expected: there's one long trace that starts with the first listener and ends when all listeners are done, and lots of short traces for all listeners after the first.

File: media.go [hunk 6]
```diff
@@ -240,6 +256,10 @@ func (pc *producerConsumer[T, U]) start() {
 			}
 
 			func() {
+				var doReadSpan *trace.Span
+				startLocalCtx, doReadSpan = trace.StartSpan(startLocalCtx, "gostream::producerConsumer (anonymous function to read)")
+				defer doReadSpan.End()
```
Comment (bazile-clyde) [inline media.go]: This can go in the defer function just below this on line 267.

File: media.go [hunk 7]
```diff
@@ -255,7 +275,7 @@ func (pc *producerConsumer[T, U]) start() {
 				} else {
 					first = false
 				}
-				media, release, err := pc.readWrapper.Read(pc.cancelCtx)
+				media, release, err := pc.readWrapper.Read(startLocalCtx)
```
Comment (bazile-clyde) [inline media.go:280]: Can we also put one more span around this _readWrapper_? That context isn't actually used anywhere and we have plans to change that but for now, it'd be great to see exactly how long this one line takes. This read usually goes deep into mediadevices so that data would be helpful. You could just put this in an anonymous function to capture any panics.

File: media.go [hunk 8]
```diff
@@ -313,6 +336,9 @@ func (pc *producerConsumer[T, U]) stop() {
 }
 
 func (pc *producerConsumer[T, U]) stopOne() {
+	_, span := trace.StartSpan(pc.cancelCtx, "gostream::producerConsumer::stopOne")
+	defer span.End()
```
Comment (bazile-clyde) [inline media.go:344]: So I'm noticing `stop` calls called by this function aren't nested under this `stopOne` call. Is there any way to make that work? Could be helpful in debugging. 

I would look at the documentation for `context.WithCancel` to make sure the context created by this function call can replace `pc.cancelCtx` without screwing up preemption. If it does get screwed up you'll see noticeable hangs and leaked goroutines when you stop this stream or at a higher level, simply stop the Viam server.

File: media.go [hunk 9]
```diff
@@ -377,6 +406,9 @@ type mediaStream[T any, U any] struct {
 }
 
 func (ms *mediaStream[T, U]) Next(ctx context.Context) (T, func(), error) {
+	ctx, nextSpan := trace.StartSpan(ctx, "gostream::mediaStream::Next")
```
Comment (bazile-clyde) [inline media.go:413]: I'd also add a span to `waitForNext` below on line 440.

File: media.go [hunk 10]
```diff
@@ -433,6 +465,9 @@ func (ms *mediaStream[T, U]) Next(ctx context.Context) (T, func(), error) {
 		}
 	}
 
+	ctx, prodConLockSpan := trace.StartSpan(ctx, "gostream::mediaStream::Next (waiting for ms.prodCon lock)")
+	defer prodConLockSpan.End()
+
```
Comment (bazile-clyde) [inline media.go:477]: You could also add a span to `Close` on line 483. I also have the same thoughts about nesting `stopOne` in `Close` the same way `stop` should be nested in `stopOne`

**Review by bazile-clyde — COMMENTED** [review]
Looks good so far. Just a few minor comments.

File: media.go (hunk 10 above)
Comment (RyanWallon) [reply media.go:477]: [Here's a trace that shows the Close -> stopOne -> stop hierarchy that you were looking for](https://console.cloud.google.com/traces/list?project=av-canary&tid=cfd45d1add23bdd55b638687640e2417&pageState=(%22traceIntervalPicker%22:(%22groupValue%22:%22PT1H%22,%22customValue%22:null))&start=1689862283893&end=1689865883893). I tried to make sure that preemption wouldn't be affected by using the original cancelCtx instead of creating a new context outright. I tested that preemption is preserved by running the same test I used to generate the trace and halting execution with a sigint, and it looked like execution stopped as quickly as it did without the changes. Let me know if that's not quite what you were looking for or if there's a better way to test these changes

File: codec/x264/encoder.go [hunk 11]
```diff
@@ -56,7 +56,7 @@ func (v *encoder) Read() (img image.Image, release func(), err error) {
 }
 
 // Encode asks the codec to process the given image.
-func (v *encoder) Encode(_ context.Context, img image.Image) ([]byte, error) {
+func (v *encoder) Encode(ctx context.Context, img image.Image) ([]byte, error) {
```
Comment (bazile-clyde) [inline codec/x264/encoder.go]: Why this change? It still looks like `ctx` isn't used anywhere.

File: media.go [hunk 12]
```diff
@@ -313,6 +340,15 @@ func (pc *producerConsumer[T, U]) stop() {
 }
 
 func (pc *producerConsumer[T, U]) stopOne() {
+	// We want to create a new span that lasts the scope of this function and no
+	// further, so we have to modify the cancelCtx and then revert it at the end
+	// of the function to what it was before.
+	originalCancelCtx := pc.cancelCtx
```
Comment (bazile-clyde) [inline media.go]: Nice! I didn't think to try using a temp variable to change and revert the context like this. Tried it out [here](https://go.dev/play/p/VQgmNw0mvGZ), and it works!

I'm curious about what happens if you remove the `originalCancelCtx`. I'd guess that the span would start on line 348 and then stop at the end of the function. That seems similar to what's going on now. Is that not the case or is there another error you ran into?

File: codec/x264/encoder.go (hunk 11 above)
Comment (RyanWallon) [reply codec/x264/encoder.go]: Good catch, I forgot to revert that change

File: media.go (hunk 12 above)
Comment (RyanWallon) [reply media.go]: Now that I look at it again you're definitely right, I wouldn't have to add the temp variable there

File: media.go [hunk 13]
```diff
@@ -501,6 +554,10 @@ func (ms *mediaSource[T, U]) Stream(ctx context.Context, errHandlers ...ErrorHan
 	prodCon.stateMu.Lock()
 	defer prodCon.stateMu.Unlock()
 
+	if currentSpan := trace.FromContext(ctx); currentSpan != nil {
+		prodCon.cancelCtx = trace.NewContext(prodCon.cancelCtx, currentSpan)
+	}
+
```
Comment (bazile-clyde) [inline media.go:556]: Why is this needed? Why not just `StartSpan` at the top of this function?

I'm also not seeing any info in our trace logs from this. If we get this working we'd also want to add a span for its close function.

File: media.go [hunk 14]
```diff
@@ -446,6 +488,17 @@ func (ms *mediaStream[T, U]) Next(ctx context.Context) (T, func(), error) {
 }
 
 func (ms *mediaStream[T, U]) Close(ctx context.Context) error {
+	// We want to create a new span that lasts the scope of this function and no
+	// further, so we have to modify the cancelCtx and then revert it at the end
+	// of the function to what it was before.
+	originalCancelCtx := ms.prodCon.cancelCtx
+	parentSpan := trace.FromContext(ctx)
+	ms.prodCon.cancelCtx = trace.NewContext(ms.prodCon.cancelCtx, parentSpan)
```
Comment (bazile-clyde) [inline media.go]: Why this construct of `StartSpan`? It looks like `FromContext` may return `nil` if there's no span in `ctx`. It may then pass `nil` (assuming `parentSpan == nil`) to `NewContext` which will map the `nil` value to a key. You can test how this would behave by trying `ms.prodCon.cancelCtx = trace.NewContext(ms.prodCon.cancelCtx, nil)`. Ideally we'd create a new span if one doesn't already exists which is what `StartSpan` does.

File: media.go [hunk 15]
```diff
@@ -446,6 +488,17 @@ func (ms *mediaStream[T, U]) Next(ctx context.Context) (T, func(), error) {
 }
 
 func (ms *mediaStream[T, U]) Close(ctx context.Context) error {
+	// We want to create a new span that lasts the scope of this function and no
+	// further, so we have to modify the cancelCtx and then revert it at the end
+	// of the function to what it was before.
+	originalCancelCtx := ms.prodCon.cancelCtx
```
Comment (bazile-clyde) [inline media.go]: This shouldn't be needed, right? Same as in the other comment.

**Review by bazile-clyde — CHANGES_REQUESTED** [review]

File: media.go (hunk 15 above)
Comment (RyanWallon) [reply media.go]: Yes, I think I had previously tried a change that would have messed up the preemption if I didn't revert the context and forgot that it doesn't matter when I'm just manipulating spans.

File: media.go (hunk 14 above)
Comment (RyanWallon) [reply media.go]: The idea is that there's a span in `ctx` which is the parent of the new span, so we take it and make it `ms.prodCon.cancelCtx`'s current span and then call `StartSpan` on it to create the child. If I'm understanding correctly, your suggestion would be `ms.prodCon.cancelCtx = trace.StartSpan(ms.prodCon.cancelCtx, "gostream::mediastream::Close")`, but then the span from `ctx` isn't the parent. I was assuming that `ctx` would always hold a span since it's called from `ReadMedia()` which itself puts a span in the context, but you're right that it's important to make sure the span isn't nil first. I can make that change

File: media.go (hunk 13 above)
Comment (RyanWallon) [reply media.go:556]: This is a similar story to the last one, we needed the span from `ctx` to be put into `prodCon.cancelCtx` so it could be used in functions which don't accept a `context.Context` but do use `prodCon.cancelCtx`. This construction wasn't meant to start a span, just pass the current one through, which is why it doesn't show up in the trace logs. I put it so low in the function because it needed to be after `prodCon` was given its `cancelCtx` and I assumed the `prodCon.stateMu` concerns the `cancelCtx` as well

Comment (bazile-clyde) [reply media.go:556]: Oh, I see. Thanks for the explanation. I was going to suggest adding a span to time this method, but there are three different contexts in this function, so figuring out which to use and how may be a bit more hassle than it's worth especially considering this may be refactored soon. I like what you've done here. This seems like the simplest solution for now. LGTM!!

**Review by bazile-clyde — APPROVED** [review]

**Comment by bazile-clyde** [comment]
@RyanBoring Great work on this! The information we gather from these additions will help us tremendously with future performance and debugging problems. LGTM! Feel free to merge.
